rfd/0137-unified-resource-view.md

```diff
@@ -0,0 +1,121 @@
+
+
+---
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 137 - Unified Resource view for the web UI
+
+## What
+
+A new view on the web UI (and eventually Connect) that will display all available resources such as Nodes, Databases, Apps, etc, in a single unified way.
+
+![Screenshot 2023-07-25 at 4 18 08 PM](https://github.com/gravitational/teleport/assets/5201977/6953289d-ec3e-4eae-bdb9-10f3cfebfd7f)
+
+## Why
+
+Rather than separating all resources into their own individual pages, a Unified Resource page would help consolidate every kind of resource into a single, searchable view. This will help with discoverability and make searching across resources a lot better. This will also lay the foundation for a much more helpful "Dashboard" type page where we can have things such as Pinned Resources (coming right after this).
+
+## Details
+
+In order to get a searchable and filterable list of all resource kinds, we will create a new, in-memory store that will initialize on auth service start and only live as long as the service is up. The watcher will update the store any time it receives an event for any of the watched kinds. We will use a very similar setup for this as our in-memory backend, but instead of storing the marshaled (unmarshaled? i forget which. TODO LEARN THE RIGHT WORD BEFORE PUBLISH) resource, we can store the actual resource. 
+
+Once that exists, we can add a new Kind `KindUnifiedResource`. This isn't an actual Kind but more of a "meta" kind that will be used to distinguish request types in `ListResources`. Using the existing pathways that `ListResources` uses will give us almost all the functionality we already need such as pagination, filtering, etc. 
+
+### New Types
+`KindUnifiedResource` will be created and used as the `RequestType` for Unified Resource requests coming from the web. This will inform `ListResources` to pull from our Unified Resource collection.
+
+Most of the core implementation is reusing logic from existing components with slight changes where needed. Our new Unified Resource store will be a b-tree similar to the backend memory structure, but the item is slightly modified to have it's `Value` be a common interface for the resources we want to track.
+```go
+type Item struct {
+	// Key is a key of the key value item
+	Key []byte
+	// Value represents a resource that is available for a unified resource request
+	Value types.ResourceWithLabels
+}
+```
+The key is in the format `name/type` to prevent name collisions across types. However, this makes the default sort a little weird where resource with UUID names will come first, even though they'll be displayed with a "friendly" name. Such as `Servers` being stored by a UUID but making the `Hostname` visible. This can change if we find a better way "on the fly" since it's only created when the auth service is started so, we aren't locked into anything in this regard. For now, we'll implement with `GetName()/GetKind()`
```

Comment:
`name/type` will allow us to sort as expected. If we did it `type/name` then all apps would show up first (if I'm understanding that correctly).

Reply:
The problem is that we need to apply sorting after all items are retrieved from the watcher anyway. So whether we use this or another type of key doesn't really matter. I think that using a BTree here will only pay off after we store items and search indices separately.

Reply:
Any change to the sort type/direction, or change to the search query/filters/etc will result in a reset of the search. essentially, its back to "page 1", similar to pagination. sorting in this case will inform the order of results that you can then infinitely scroll through. Our infinite scroll is really just pagination behind the scenes, but only "going to next page" and appending the results.

Reply:
@avatus Could you look once more to my comment above? Since we ended up sorting in the API layer anyway, it's now just an arbitrary decision that doesn't have anything to do with sorting. I'd remove this justification from here.